New nullability annotations on custom type adapters break nullable types

In 0.4.2 - nullable fields backed by custom types will break because the CustomTypeAdapter API only allows for returning nonnull values. in loosely checked Java code this flies, but in kotlin it becomes either a compile error or runtime error (if it is null and you used `!!`).

Will link a repro branch tomorrow but putting up for visibility now. I think either the return nullability should be relaxed or a nullsafe wrapper should be introduced and used in the event of nullable types. The former is probably more flexible, as the only way the latter works is if the nullsafe adapter try/catches the delegate adapter

Here's a repro branch: https://github.com/hzsweers/CatchUp/tree/z/apollo042

The generated code issue is in `GithubSearchQuery$AsRepository$Mapper#map`, this line in particular

```java
final HttpUrl homepageUrl = reader.readCustomType((ResponseField.CustomTypeField) $responseFields[4]);
```

This doesn't handle the nullability of the `homepageUrl` field, and delegates directly to the `HttpUrlAdapter` implementation, which must nullcheck now in order to compile

```kotlin
class HttpUrlApolloAdapter : CustomTypeAdapter<HttpUrl> {
  override fun decode(s: String) = HttpUrl.parse(s)!!

  override fun encode(o: HttpUrl) = o.toString()
}
```

This results in the following trace at runtime during deserialization if that field is null

```
10-21 04:00:46.340 2603-2603/io.sweers.catchup.debug E/ServiceController$loadD: HttpException
                                                                                com.apollographql.apollo.exception.ApolloParseException: Failed to parse http response
                                                                                    at com.apollographql.apollo.internal.interceptor.ApolloParseInterceptor.parse(ApolloParseInterceptor.java:101)
                                                                                    at com.apollographql.apollo.internal.interceptor.ApolloParseInterceptor.access$100(ApolloParseInterceptor.java:28)
                                                                                    at com.apollographql.apollo.internal.interceptor.ApolloParseInterceptor$1.onResponse(ApolloParseInterceptor.java:53)
                                                                                    at com.apollographql.apollo.internal.interceptor.ApolloServerInterceptor$1$1.onResponse(ApolloServerInterceptor.java:93)
                                                                                    at com.google.firebase.perf.network.zzg.onResponse(Unknown Source:16)
                                                                                    at okhttp3.RealCall$AsyncCall.execute(RealCall.java:153)
                                                                                    at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
                                                                                    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
                                                                                    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
                                                                                    at java.lang.Thread.run(Thread.java:764)
                                                                                 Caused by: kotlin.KotlinNullPointerException
                                                                                    at io.sweers.catchup.data.HttpUrlApolloAdapter.decode(HttpUrlApolloAdapter.kt:26)
                                                                                    at io.sweers.catchup.data.HttpUrlApolloAdapter.decode(HttpUrlApolloAdapter.kt:25)
                                                                                    at com.apollographql.apollo.internal.response.RealResponseReader.readCustomType(RealResponseReader.java:189)
                                                                                    at io.sweers.catchup.service.github.GitHubSearchQuery$AsRepository$Mapper.map(GitHubSearchQuery.java:920)
                                                                                    at io.sweers.catchup.service.github.GitHubSearchQuery$Node$Mapper$1.read(GitHubSearchQuery.java:667)
                                                                                    at io.sweers.catchup.service.github.GitHubSearchQuery$Node$Mapper$1.read(GitHubSearchQuery.java:664)
                                                                                    at com.apollographql.apollo.internal.response.RealResponseReader.readConditional(RealResponseReader.java:215)
                                                                                    at io.sweers.catchup.service.github.GitHubSearchQuery$Node$Mapper.map(GitHubSearchQuery.java:664)
                                                                                    at io.sweers.catchup.service.github.GitHubSearchQuery$Search$Mapper$2$1.read(GitHubSearchQuery.java:461)
                                                                                    at io.sweers.catchup.service.github.GitHubSearchQuery$Search$Mapper$2$1.read(GitHubSearchQuery.java:458)
                                                                                    at com.apollographql.apollo.internal.response.RealResponseReader$ListItemReader.readObject(RealResponseReader.java:306)
                                                                                    at io.sweers.catchup.service.github.GitHubSearchQuery$Search$Mapper$2.read(GitHubSearchQuery.java:458)
                                                                                    at io.sweers.catchup.service.github.GitHubSearchQuery$Search$Mapper$2.read(GitHubSearchQuery.java:455)
                                                                                    at com.apollographql.apollo.internal.response.RealResponseReader.readList(RealResponseReader.java:163)
                                                                                    at io.sweers.catchup.service.github.GitHubSearchQuery$Search$Mapper.map(GitHubSearchQuery.java:455)
                                                                                    at io.sweers.catchup.service.github.GitHubSearchQuery$Data$Mapper$1.read(GitHubSearchQuery.java:314)
                                                                                    at io.sweers.catchup.service.github.GitHubSearchQuery$Data$Mapper$1.read(GitHubSearchQuery.java:311)
                                                                                    at com.apollographql.apollo.internal.response.RealResponseReader.readObject(RealResponseReader.java:136)
                                                                                    at io.sweers.catchup.service.github.GitHubSearchQuery$Data$Mapper.map(GitHubSearchQuery.java:311)
                                                                                    at io.sweers.catchup.service.github.GitHubSearchQuery$Data$Mapper.map(GitHubSearchQuery.java:306)
                                                                                    at com.apollographql.apollo.internal.response.OperationResponseParser$1.read(OperationResponseParser.java:56)
                                                                                    at com.apollographql.apollo.internal.json.ResponseJsonStreamReader.nextObject(ResponseJsonStreamReader.java:86)
                                                                                    at com.apollographql.apollo.internal.response.OperationResponseParser.parse(OperationResponseParser.java:51)
                                                                                    at com.apollographql.apollo.internal.interceptor.ApolloParseInterceptor.parse(ApolloParseInterceptor.java:87)
                                                                                    at com.apollographql.apollo.internal.interceptor.ApolloParseInterceptor.access$100(ApolloParseInterceptor.java:28) 
                                                                                    at com.apollographql.apollo.internal.interceptor.ApolloParseInterceptor$1.onResponse(ApolloParseInterceptor.java:53) 
                                                                                    at com.apollographql.apollo.internal.interceptor.ApolloServerInterceptor$1$1.onResponse(ApolloServerInterceptor.java:93) 
                                                                                    at com.google.firebase.perf.network.zzg.onResponse(Unknown Source:16) 
                                                                                    at okhttp3.RealCall$AsyncCall.execute(RealCall.java:153) 
                                                                                    at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32) 
                                                                                    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162) 
                                                                                    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636) 
                                                                                    at java.lang.Thread.run(Thread.java:764) 
```

--- a/apollo-runtime/src/main/java/com/apollographql/apollo/CustomTypeAdapter.java
+++ b/apollo-runtime/src/main/java/com/apollographql/apollo/CustomTypeAdapter.java
@@ -67,7 +67,7 @@ public interface CustomTypeAdapter<T> {
    * @param value the value to de-serialize
    * @return custom scalar type
    */
-  @Nonnull T decode(@Nonnull String value);
+  T decode(@Nonnull String value);
 
   /**
    * Serializes the custom scalar type to the corresponding string value. Usually used in serializing variables or input

--- a/apollo-runtime/src/main/java/com/apollographql/apollo/internal/response/RealResponseReader.java
+++ b/apollo-runtime/src/main/java/com/apollographql/apollo/internal/response/RealResponseReader.java
@@ -161,7 +161,9 @@ import java.util.Map;
         Object value = values.get(i);
         if (value != null) {
           T item = (T) listReader.read(new ListItemReader(field, value));
-          result.add(item);
+          if (item != null) {
+            result.add(item);
+          }
         }
         readerShadow.didResolveElement(i);
       }
@@ -185,8 +187,9 @@ import java.util.Map;
       result = null;
     } else {
       CustomTypeAdapter<T> typeAdapter = scalarTypeAdapters.adapterFor(field.scalarType());
-      readerShadow.didResolveScalar(value);
       result = typeAdapter.decode(value.toString());
+      checkValue(field, result);
+      readerShadow.didResolveScalar(value);
     }
     didResolve(field);
     return result;

--- a/apollo-runtime/src/test/java/com/apollographql/apollo/internal/reader/ResponseReaderTest.java
+++ b/apollo-runtime/src/test/java/com/apollographql/apollo/internal/reader/ResponseReaderTest.java
@@ -1,5 +1,7 @@
 package com.apollographql.apollo.internal.reader;
 
+import com.google.common.truth.Truth;
+
 import com.apollographql.apollo.CustomTypeAdapter;
 import com.apollographql.apollo.api.Operation;
 import com.apollographql.apollo.api.OperationName;
@@ -19,6 +21,7 @@ import com.apollographql.apollo.internal.response.ScalarTypeAdapters;
 import org.junit.Test;
 
 import java.math.BigDecimal;
+import java.net.URL;
 import java.text.ParseException;
 import java.text.SimpleDateFormat;
 import java.util.Collection;
@@ -170,9 +173,9 @@ public class ResponseReaderTest {
   }
 
   @Test public void readCustom() throws Exception {
-    ResponseField successField = ResponseField.forCustomType("successFieldResponseName", "successFieldName", null, false, CUSTOM_TYPE, NO_CONDITIONS);
+    ResponseField successField = ResponseField.forCustomType("successFieldResponseName", "successFieldName", null, false, DATE_CUSTOM_TYPE, NO_CONDITIONS);
     ResponseField classCastExceptionField = ResponseField.forCustomType("classCastExceptionFieldResponseName", "classCastExceptionFieldName",
-        null, false, CUSTOM_TYPE, NO_CONDITIONS);
+        null, false, DATE_CUSTOM_TYPE, NO_CONDITIONS);
 
     Map<String, Object> recordSet = new HashMap<>();
     recordSet.put("successFieldResponseName", "2017-04-16");
@@ -187,6 +190,23 @@ public class ResponseReaderTest {
     } catch (ClassCastException expected) {
       // expected
     }
+  }
+
+  @Test public void readCustomWithDecodedNullValue() throws Exception {
+    Map<String, Object> recordSet = new HashMap<>();
+    recordSet.put("responseName", "http:://");
+    RealResponseReader<Map<String, Object>> responseReader = responseReader(recordSet);
+
+    ResponseField field = ResponseField.forCustomType("responseName", "fieldName", null, false, URL_CUSTOM_TYPE, NO_CONDITIONS);
+    try {
+      responseReader.readCustomType((ResponseField.CustomTypeField) field);
+      fail("expected NullPointerException");
+    } catch (NullPointerException e) {
+      // expected
+    }
+
+    field = ResponseField.forCustomType("responseName", "fieldName", null, true, URL_CUSTOM_TYPE, NO_CONDITIONS);
+    Truth.assertThat(responseReader.readCustomType((ResponseField.CustomTypeField) field)).isNull();
   }
 
   @Test public void readCustomWithDefaultAdapter() throws Exception {
@@ -225,7 +245,7 @@ public class ResponseReaderTest {
     try {
       responseReader.readCustomType(unsupportedField);
       fail("Expect IllegalArgumentException");
-    } catch (IllegalArgumentException expected){
+    } catch (IllegalArgumentException expected) {
       // expected
     }
   }
@@ -428,7 +448,7 @@ public class ResponseReaderTest {
     RealResponseReader<Map<String, Object>> responseReader = responseReader(recordSet);
     assertThat(responseReader.readList(successField, new ResponseReader.ListReader() {
       @Override public Object read(ResponseReader.ListItemReader reader) {
-        return reader.readCustomType(CUSTOM_TYPE);
+        return reader.readCustomType(DATE_CUSTOM_TYPE);
       }
     })).isEqualTo(asList(DATE_TIME_FORMAT.parse("2017-04-16"), DATE_TIME_FORMAT.parse("2017-04-17"), DATE_TIME_FORMAT.parse("2017-04-18")));
 
@@ -442,6 +462,19 @@ public class ResponseReaderTest {
     } catch (ClassCastException expected) {
       // expected
     }
+  }
+
+  @Test public void readCustomListWithDecodedNull() throws Exception {
+    final Map<String, Object> recordSet = new HashMap<>();
+    recordSet.put("responseName", asList("http://", "http://"));
+    RealResponseReader<Map<String, Object>> responseReader = responseReader(recordSet);
+
+    ResponseField field = ResponseField.forList("responseName", "fieldName", null, false, NO_CONDITIONS);
+    Truth.assertThat(responseReader.readList(field, new ResponseReader.ListReader() {
+      @Override public Object read(ResponseReader.ListItemReader reader) {
+        return reader.readCustomType(URL_CUSTOM_TYPE);
+      }
+    })).isEmpty();
   }
 
   @Test public void readObjectList() throws Exception {
@@ -538,7 +571,7 @@ public class ResponseReaderTest {
       }
     });
     responseReader.readCustomType((ResponseField.CustomTypeField) ResponseField.forCustomType("customTypeField", "customTypeField",
-        null, true, CUSTOM_TYPE, NO_CONDITIONS));
+        null, true, DATE_CUSTOM_TYPE, NO_CONDITIONS));
   }
 
   @Test public void mandatoryFieldsIOException() throws Exception {
@@ -603,7 +636,7 @@ public class ResponseReaderTest {
 
     try {
       responseReader.readCustomType((ResponseField.CustomTypeField) ResponseField.forCustomType("customTypeField",
-          "customTypeField", null, false, CUSTOM_TYPE, NO_CONDITIONS));
+          "customTypeField", null, false, DATE_CUSTOM_TYPE, NO_CONDITIONS));
       fail("expected NullPointerException");
     } catch (NullPointerException expected) {
       //expected
@@ -670,13 +703,22 @@ public class ResponseReaderTest {
   @SuppressWarnings("unchecked") private static RealResponseReader<Map<String, Object>> responseReader(
       Map<String, Object> recordSet) {
     Map<ScalarType, CustomTypeAdapter> customTypeAdapters = new HashMap<>();
-    customTypeAdapters.put(CUSTOM_TYPE, new CustomTypeAdapter() {
+    customTypeAdapters.put(DATE_CUSTOM_TYPE, new CustomTypeAdapter() {
       @Override public Object decode(String value) {
         try {
           return DATE_TIME_FORMAT.parse(value);
         } catch (ParseException e) {
           throw new ClassCastException();
         }
+      }
+
+      @Override public String encode(Object value) {
+        return null;
+      }
+    });
+    customTypeAdapters.put(URL_CUSTOM_TYPE, new CustomTypeAdapter() {
+      @Override public Object decode(String value) {
+        return null;
       }
 
       @Override public String encode(Object value) {
@@ -699,13 +741,23 @@ public class ResponseReaderTest {
     };
   }
 
-  private static final ScalarType CUSTOM_TYPE = new ScalarType() {
+  private static final ScalarType DATE_CUSTOM_TYPE = new ScalarType() {
     @Override public String typeName() {
       return Date.class.getName();
     }
 
     @Override public Class javaType() {
       return Date.class;
+    }
+  };
+
+  private static final ScalarType URL_CUSTOM_TYPE = new ScalarType() {
+    @Override public String typeName() {
+      return URL.class.getName();
+    }
+
+    @Override public Class javaType() {
+      return URL.class;
     }
   };
 

